### PR TITLE
Fix boss score, add boss slayer

### PR DESCRIPTION
**Pull request by direwolfclaw**
### Brief description of the changes:
--Fixed bugs associated with boss kill scoring.
--Spider vehicle kills are now counted toward score.
--Added bonus score for boss kills (implemented as ratio of boss HP to base class)
--Added new MVP title "Boss Slayer" for killing the most bosses.

### Tested Changes:
- [x] I've tested the changes locally or with people.
- [ ] I've not tested the changes.

### Known issues:
--Only issue I can find is the score menu in comfy panel seems to have a max of five columns visible, even if Ncolumns is now 6.

### Vote resulst : 
image
https://discord.com/channels/823696400797138974/823771211421974579/1161763825574498315

### CODE REVIEW

**Comment by direwolfclaw** [comment]
also this deprecates #366

**Comment by Ragnarok77-factorio** [comment]
Added the vote result to the PR.

**Comment by XVhc6A** [comment]
> Only issue I can find is the score menu in comfy panel seems to have a max of five columns visible, even if Ncolumns is now 6.

This is because of the width of the text boxes. You either should reduce their size (currently 150) or increase the width of the panel.

I ran this locally for a bit,  didn't have any issues and saw boss kills change when I killed bosses.

**Comment by direwolfclaw** [comment]
I tried to manually fix the conflicts (replied to your email about this) - even though I manually changed the offending lines to match the current code, it didn't work. I will gently point out that when submitted there were no conflicts xd. How do we fix this? Do I just make a new PR forking from the current code?

And yes, XVHc I did some more work on BBB and figured out how to change the size of the whole panel - are people OK with this? Otherwise we can make the columns narrow, but they might not be legible. Moreover, if we add all of the MVPs voted for, there is no way the pane will be wide enough regardless. Maybe we just enable horizontal scrolling instead?

I had also asked do you want me to code the other MVPs so we can add them all together? Waller is trivial. First sender is fairly straight forward. Builder for each hour will be slightly challenging (my thought is to turn it into a list and use floor of ticks_played / 3600*60 to figure out which hour to add to). Should builder now ignore wall? I also think builder should ignore belt, but this is mostly because I seem to get builder more often than I deserve it, based only on building huge numbers of belts lol.

**Comment by direwolfclaw** [comment]
Finally, would we prefer aggregate damage done to bosses? It's simple to change. In some ways I agree with soldier it will more accurately reflect contributions to winning the game, even if less satisfying.

**Comment by XVhc6A** [comment]
> I tried to manually fix the conflicts (replied to your email about this) - even though I manually changed the offending lines to match the current code, it didn't work. I will gently point out that when submitted there were no conflicts xd. How do we fix this (the conflicts are now trivial)? Do I just make a new PR forking from the current code?

Looks like you got it, but no -- the github UI should help you with the process : )

but yes, sorry that we merged other things, we can also do the conflict merge for you, but some people on github have pride and stuff so it's usually courtesy to ask the Author first.

> And yes, XVHc I did some more work on BBB and figured out how to change the size of the whole panel - are people OK with this? Otherwise we can make the columns narrow, but they might not be legible. Moreover, if we add all of the MVPs voted for, there is no way the pane will be wide enough regardless. Maybe we just enable horizontal scrolling instead?

I think we can go 1-2 columns wider, and reduce column width by like 15-25? but after this we should do horizontal scrolling.  If you want to try implementing horizontal scrolling  - have at it.  I seem to remember some weirdness with having both horizontal and vertical scrollbars, but maybe I'm thinking of something else

just upload (paste) a screenshot with that you decide.

> I had also asked do you want me to code the other MVPs so we can add them all together? Waller is trivial. First sender is fairly straight forward. Builder for each hour will be slightly challenging (my thought is to turn it into a list and use floor of ticks_played / 3600*60 to figure out which hour to add to). Should builder now ignore wall? I also think builder should ignore belt, but this is mostly because I seem to get builder more often than I deserve it, based only on building huge numbers of belts lol.

Just do waller / sender with this PR, for the /hr thing do that in a separate PR. 

I don't think it's necessary to exclude anything from builder, but I'd remove furnace, belts, and walls if you do. Try not to overthink it : ) -- it is just for fun, if someone takes offense that you're the best builder in the game, they can submit their own PR.

**Comment by direwolfclaw** [comment]
i'm going to consider this done and leave the rest for another day. what needs to happen is to merge/squash all of the commits into one. not sure how to do that here but i see how to do it in git itself. this is quite simple and self-contained.

**Comment by XVhc6A** [comment]
> needs to happen is to merge/squash all of the commits into one

You can do it yourself, or we can do it when we merge it (github has an option for that when we click merge)

File: functions/boss_unit.lua [hunk 1]
```diff
@@ -35,8 +35,8 @@ local function on_entity_damaged(event)
 	entity.health = entity.health + event.final_damage_amount
 	boss.health = boss.health - event.final_damage_amount
 	if boss.health <= 0 then
+		entity.die(event.cause.force, event.cause)
 		global.boss_units[entity.unit_number] = nil
```
Comment (XVhc6A) [inline functions/boss_unit.lua:39]:
```suggestion
		global.boss_units[entity.unit_number] = nil
		entity.die(event.cause.force, event.cause)
```

I get the error: 
```
LuaEntity API call when LuaEntity was invalid. in .../factorio/temp/currently-playing/functions/boss_unit.lua:31
```
when bosses die right now, You should reverse this so the entity dies after you get its unit_number.

**Comment by XVhc6A** [comment]
Looks good! I see one error but I think it's an easy fix. (was locally for me)

Once the above suggested change is approved  I belive this can be merged


image

**Review by XVhc6A — APPROVED** [review]
see comment

**Pull request closed without merging**